### PR TITLE
BUG: spatial.cKDTree: do not slide the median

### LINKED ISSUE #20605: BUG: Sliding is applied when using median as pivot in cKDTree, should only be used for midpoint
As the title says…

The bug relates to how the kd-tree is constructed and does not produce a logic error.

However, sliding midpoint is sliding midpoint. Median is median. Sliding median is not a thing.

This bug has been fixed before, but has reappeared (regression due to #14355).

@peterbell10

### PR DESCRIPTION
**Pull request by sturlamolden**
Never slide the pivot when using median as pivot. Only slide the pivot when pivot is the midpoint. That’s what sliding midpoint means.

Sliding median is not a thing. A median is always the median.

Closes #20605

### CODE REVIEW

**Review by j-bowhay — COMMENTED** [review]
Is there a regression test that can be added so this doesn't get accidentally broken again or is the change buried too deep?

**Review by tylerjereddy — COMMENTED** [review]
While looking for a regression test case, I seem to have found one that crashes on this PR branch and doesn't crash on `main`:

```diff
--- a/scipy/spatial/tests/test_kdtree.py
+++ b/scipy/spatial/tests/test_kdtree.py
@@ -1531,3 +1531,13 @@ def test_gh_18800(incantation):
     tree = incantation(points, 10)
     tree.query(arr_like, 1)
     tree.query_ball_point(arr_like, 200)
+
+
+def test_gh_20605():
+    # this currently crashes in gh-20606
+    # but not on main
+    data = np.full((100, 2), 4)
+    data[:50, :] = 5
+    data[52:60, 1] = 8
+    balanced_tree = KDTree(data=data,
+                           balanced_tree=True)
```

I suppose that makes me a little nervous.

**Comment by sturlamolden** [comment]
It seems when splitting on the median, we can get a situation where all points end up in the same branch. This will then trigger an infinite recursion, resulting in a stack overflow.

**Comment by sturlamolden** [comment]
I also find this strange code in the kd-tree construction:
```C++
if (CKDTREE_UNLIKELY(p == start_idx || p == end_idx)) {
            // All children are equal in this dimension, try again with new bounds
            assert(!_compact);
            self->tree_buffer->pop_back();
            std::vector<double> tmp_bounds(2 * m);
            double* tmp_mins = &tmp_bounds[0];
            std::copy_n(mins, m, tmp_mins);
            double* tmp_maxes = &tmp_bounds[m];
            std::copy_n(maxes, m, tmp_maxes);

            const auto fixed_val = data[indices[start_idx]*m + d];
            tmp_mins[d] = fixed_val;
            tmp_maxes[d] = fixed_val;

            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact);
        }
```
It tries to solve the issue of empty branches by pretending the spread along a dimension is zero. However, it propagates a false max value into the kd-tree, which can result in bogus query results later on. It came in with d13a48c and is a logic error.

**Comment by sturlamolden** [comment]
> While looking for a regression test case, I seem to have found one that crashes on this PR branch and doesn't crash on `main`:
> 
> ```diff
> --- a/scipy/spatial/tests/test_kdtree.py
> +++ b/scipy/spatial/tests/test_kdtree.py
> @@ -1531,3 +1531,13 @@ def test_gh_18800(incantation):
>      tree = incantation(points, 10)
>      tree.query(arr_like, 1)
>      tree.query_ball_point(arr_like, 200)
> +
> +
> +def test_gh_20605():
> +    # this currently crashes in gh-20606
> +    # but not on main
> +    data = np.full((100, 2), 4)
> +    data[:50, :] = 5
> +    data[52:60, 1] = 8
> +    balanced_tree = KDTree(data=data,
> +                           balanced_tree=True)
> ```
> 
> I suppose that makes me a little nervous.

Thanks for the test. It revealed two more bugs in the new C++ code.

There are also stability issues with using the median as pivot I have not seen before. 

This makes me a bit nervous too.

**Comment by tylerjereddy** [comment]
@sturlamolden is this PR in a state where you think we should consider merging it before I branch for SciPy `1.14.0`? It looks like you added my test and CI is green.

File: scipy/spatial/ckdtree/src/build.cxx [hunk 1]
```diff
@@ -112,53 +120,50 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
             auto mid = node_indices + n_points / 2;
             std::nth_element(
                 node_indices, mid, node_indices + n_points, index_compare);
-
             split = data[*mid * m + d];
-            p = partition_pivot(node_indices, mid, split);
+            p = partition_pivot(indices + start_idx, indices + end_idx, split);
```
Comment (peterbell10) [inline scipy/spatial/ckdtree/src/build.cxx:124]: Why change this? We know the range `[node_indices, mid)` is partitioned by `nth_element` such that all values less than the pivot are in that range already.

File: scipy/spatial/ckdtree/src/build.cxx [hunk 2]
```diff
@@ -112,53 +120,50 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
...
-            double* tmp_maxes = &tmp_bounds[m];
-            std::copy_n(maxes, m, tmp_maxes);
-
-            const auto fixed_val = data[indices[start_idx]*m + d];
-            tmp_mins[d] = fixed_val;
-            tmp_maxes[d] = fixed_val;
-
-            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact);
+            taboo[d] = true;
+            return build(self, start_idx, end_idx, maxes, mins, _median, _compact, taboo, taboo_depth+1);
```
Comment (peterbell10) [inline scipy/spatial/ckdtree/src/build.cxx:156]: > I also find this strange code in the kd-tree construction [...] It tries to solve the issue of empty branches by pretending the spread along a dimension is zero.

Can you describe a situation where partitioning failed once already, but the data isn't fully degenerate?

**Comment by sturlamolden** [comment]
Using the median as a pivot has a fail mode when the median is the largest value in a node. When this happen, the larger child node is empty and the lesser is equal to the parent. The the recursion repeats itself ad infinitum and crashes when it runs out of stack. Sliding the pivot prevents this from happening. Logically we must therefore always apply the sliding midpoint rule, also when using the median. We should therefore rather change the documentation to reflect what the kd-tree actually does.

**Comment by sturlamolden** [comment]
Note that the rationale here is different from the sliding midpoint rule: It exists to prevent slow queries when a node is streched along one dimension. In case of the median we want to avoid infinite recursion when building the kd-tree.

**Comment by lucascolley** [comment]
> We should therefore rather change the documentation to reflect what the kd-tree actually does.

Hey @sturlamolden, just checking in as we are approaching the 1.15.0 release process - would you like to return to this?

**Comment by lucascolley** [comment]
I've closed this PR as it sounds like the code changes here aren't relevant to the desired solution to the linked issue any more - feel free to reopen if that is not the case!

**Pull request closed without merging**